### PR TITLE
Implement string tokenization functionalities as defined in Fortran 2023

### PR DESCRIPTION
**Pull request by kuanchihwang**
### Tag name (required for release branches):

None

### Originator(s):

kuanchihwang

### Descriptions (include the issue title, and the keyword ['closes', 'fixes', 'resolves'] followed by the issue number):

This PR implements the new string tokenization functionalities that have been standardized in Fortran 2023. Specifically, two new intrinsic procedures have been defined with the following signatures:

1. `split (string, set, pos [, back])`, defined in section 16.9.196.
2. `tokenize (string, set, tokens [, separator])` or `tokenize (string, set, first, last)`, defined in section 16.9.210.

The implementations provided by this PR should be fully compliant with the specifications in Fortran 2023.

Additionally, the `parse_stream_name` function, which is part of the I/O infrastructure for MPAS dynamical core, is refactored to use the `tokenize` subroutine for better robustness.

### Describe any changes made to the build system:

None

### Describe any changes made to the namelist:

None

### List any changes to the defaults for the input datasets (e.g., boundary datasets):

None

### List all files eliminated and why:

None

### List all files added and what they do:

None

### List all existing files that have been modified, and describe the changes:

```
M       src/core_utils/string_core_utils.F90
M       src/utils/string_utils.F90
M       test/unit/fortran/src/core_utils/test_string_core_utils.pf
  * Avoid unnecessary renames from `stringify` to `core_stringify`
  * Also make `split` and `tokenize` available in `string_utils` module
M       src/dynamics/mpas/driver/dyn_mpas_procedures.F90
  * Implement `split` and `tokenize` from Fortran 2023
M       src/dynamics/mpas/driver/dyn_mpas_subdriver.F90
  * Rewrite `parse_stream_name` in terms of `tokenize`
M       src/dynamics/mpas/dyn_procedures.F90
  * Add missing dates
M       src/dynamics/mpas/tests/unit/test_dyn_mpas_procedures.pf
  * Implement unit tests for `split` and `tokenize`
```

### Regression tests:

```
  SMS_Ln9.ne3pg3_ne3pg3_mg37.FKESSLER.derecho_intel.cam-outfrq_se_cslam_multitape (Overall: NLFAIL)
```

Except for the known failing test above, all the other tests pass with respect to the last baseline, `sima0_09_000`.

### CODE REVIEW

**Comment by jimmielin** [comment]
Thanks @kuanchihwang. I wonder what made the GNU regression tests different (it looks like the snapshot runs are fine which is great!). Maybe we have to accept the new baselines as it is difficult to know what changed at this point.

`help@` hasn't gotten back to me yet about the `libhugetlbfs.so` issue I've been running to that crashes GNU runs at runtime, but it looks like the file is now available on the compute nodes again (I was able to run a case as well). Maybe they are still in the progress of fixing issues?

**Comment by jimmielin** [comment]
Hi @kuanchihwang I pushed the changed GNU baselines post-Derecho upgrade as part of `sima0_09_000` (#426) so hopefully you will have all GNU tests bit-for-bit now. Thanks!

File: src/dynamics/mpas/tests/unit/test_dyn_mpas_procedures.pf [hunk 1]
```diff
@@ -1052,4 +1060,336 @@ contains
...
+    @test
+    subroutine test_split_by_forward_searches()
+        use dyn_mpas_procedures, only: split
+        use funit
+
+        character(*), parameter :: test_string_comma_separated = &
+            ',,Mercury,Venus,,Earth,,,Mars,,Jupiter,Saturn,,Uranus,,,Neptune,,'
+        character(*), parameter :: test_set_comma = ','
+        character(*), parameter :: test_string_comma_space_separated = &
+            ', Mercury,Venus ,Earth , Mars, Jupiter,Saturn ,Uranus , Neptune, '
```
Comment (nusbaume) [inline src/dynamics/mpas/tests/unit/test_dyn_mpas_procedures.pf]: Just to make sure we have covered every edge case, should we include a separator that is just a space in the provided string?  For example:


```suggestion
            ', Mercury,Venus ,Earth , Mars, Jupiter Saturn ,Uranus , Neptune, '
```

File: src/dynamics/mpas/tests/unit/test_dyn_mpas_procedures.pf [hunk 2]
```diff
@@ -38,6 +38,14 @@ module test_dyn_mpas_procedures
     public :: test_index_unique_by_logical_arrays
     public :: test_index_unique_by_real_arrays_real32
     public :: test_index_unique_by_real_arrays_real64
+    public :: test_split_by_empty_string_set
```
[the remark this answers is not in the document]
Comment (kuanchihwang) [reply src/dynamics/mpas/tests/unit/test_dyn_mpas_procedures.pf]: Yes, added in https://github.com/ESCOMP/CAM-SIMA/pull/443/commits/6facbe101437b7694b8a9594dc7c78591268b856. GitHub Actions logs also confirm these tests pass.
